galsim/zernike.py

```diff
@@ -0,0 +1,398 @@
+        return _annular_zern_rho_coefs(2*j, 0, eps)[::2]
+    else:  # Equation (A13)
+        num = 2*(2*j+2*m-1) * _h(m-1, j, eps)
+        den = (j+m)*(1-eps**2)*_Q(m-1, j, eps)[0]
+        summation = np.zeros((j+1,), dtype=float)
+        for i in range(j+1):
+            qq = _Q(m-1, i, eps)
+            qq = qq*qq[0]  # Don't use *= here since it modifies the cache!
+            summation[:i+1] += qq/_h(m-1, i, eps)
+        return summation * num / den
+_Q = LRU_Cache(__Q)
+
+
+def __annular_zern_rho_coefs(n, m, eps):
+    """Compute coefficients of radial part of annular Zernike (n, m), with fractional linear
+    obscuration eps.
+    """
+    out = np.zeros((n+1,), dtype=float)
+    m = abs(m)
+    if m == 0:  # Equation (18)
+        norm = 1./(1-eps**2)
+        # R[n, m=0, eps](r^2) = R[n, m=0, eps=0]((r^2 - eps^2)/(1 - eps^2))
+        # Implement this by retrieving R[n, 0] coefficients of (r^2)^k and
+        # multiplying in the binomial (in r^2) expansion of ((r^2 - eps^2)/(1 - eps^2))^k
+        coefs = _zern_rho_coefs(n, 0)
+        for i, coef in enumerate(coefs):
+            if i % 2 == 1: continue
+            j = i // 2
+            more_coefs = (norm**j) * binomial(-eps**2, 1, j)
+            out[0:i+1:2] += coef*more_coefs
+    elif m == n:  # Equation (25)
+        norm = 1./np.sqrt(np.sum((eps**2)**np.arange(n+1)))
+        out[n] = norm
+    else:  # Equation (A1)
+        j = (n-m)//2
+        norm = np.sqrt((1-eps**2)/(2*(2*j+m+1) * _h(m,j,eps)))
+        out[m::2] = norm * _Q(m, j, eps)
+    return out
+_annular_zern_rho_coefs = LRU_Cache(__annular_zern_rho_coefs)
+
+
+class Zernike(object):
+    """A class to represent Zernike polynomial series
+    (http://en.wikipedia.org/wiki/Zernike_polynomials#Zernike_polynomials).
+
+    Zernike polynomials form an orthonormal basis over the unit circle.  The convention used here is
+    for the normality constant to equal the area of integration, which is pi for the unit circle.
+    I.e.,
+
+        \int_{unit circle} Z_i Z_j dA = \pi \delta_{i, j}.
+
+    Two generalizations of the unit circle Zernike polynomials are also available in this class:
+    annular Zernike polynomials, and polynomials defined over non-unit-radius circles.
+
+    Annular Zernikes are orthonormal over an annulus instead of a circle (see Mahajan, J. Opt. Soc.
+    Am. 71, 1, (1981)).  Similarly, the non-unit-radius polynomials are orthonormal over a region
+    with outer radius not equal to 1.  Taken together, these generalizations yield the
+    orthonormality condition:
+
+        \int_{annulus} Z_i Z_j dA = \pi (R_outer^2 - R_inner^2) \delta_{i, j}
+
+    where 0 < R_inner < R_outer indicate the inner and outer radii of the annulus over which the
+    polynomials are orthonormal.
+
+    The indexing convention for i and j is that from Noll, J. Opt. Soc. Am. 66, 207-2011(1976),
+    though offset by 1 because python is 0-indexed while the Noll indices start at 1.  For example,
```

Comment:
This seems potentially confusing and error-prone.  People think of Z4 as defocus, so it seems like it would be easy for people to forget to subtract 1 when using this class.  (Or when setting up their array to pass to Zernike.) 

Can we just hide the 1->0 offset internally and make the public API use the real Noll indices?  (Like we did for the `aberrations` parameter of OpticalPSF.)

Reply:
This one gave me a little pause, especially now that there's output that needs to be conforming (e.g. zernikeBasis) as well as input.  I think I probably would have preferred the opposite convention if we were starting from scratch, but since there's already a well-established convention from OpticalPSF, and it's good to stay consistent, I went ahead and hid the 1->0 offset as you suggest.